File: src/swift/Lesson6/lesson6.swift
/*
 * SPDX-FileCopyrightText: (C) 2025 a dinosaur
 * SPDX-License-Identifier: Zlib
 */

import Foundation
import SDLSwift
import NeHe
import simd

struct Lesson6: AppDelegate
{
	struct Vertex
	{
		let position: SIMD3<Float>, texcoord: SIMD2<Float>

		init(_ position: SIMD3<Float>, _ texcoord: SIMD2<Float>)
		{
			self.position = position
			self.texcoord = texcoord
		}
	}

	static let vertices =
	[
		// Front Face
		Vertex(.init(-1.0, -1.0,  1.0), .init(0.0, 0.0)),
		Vertex(.init( 1.0, -1.0,  1.0), .init(1.0, 0.0)),
		Vertex(.init( 1.0,  1.0,  1.0), .init(1.0, 1.0)),
		Vertex(.init(-1.0,  1.0,  1.0), .init(0.0, 1.0)),
		// Back Face
		Vertex(.init(-1.0, -1.0, -1.0), .init(1.0, 0.0)),
		Vertex(.init(-1.0,  1.0, -1.0), .init(1.0, 1.0)),
		Vertex(.init( 1.0,  1.0, -1.0), .init(0.0, 1.0)),
		Vertex(.init( 1.0, -1.0, -1.0), .init(0.0, 0.0)),
		// Top Face
		Vertex(.init(-1.0,  1.0, -1.0), .init(0.0, 1.0)),
		Vertex(.init(-1.0,  1.0,  1.0), .init(0.0, 0.0)),
		Vertex(.init( 1.0,  1.0,  1.0), .init(1.0, 0.0)),
		Vertex(.init( 1.0,  1.0, -1.0), .init(1.0, 1.0)),
		// Bottom Face
		Vertex(.init(-1.0, -1.0, -1.0), .init(1.0, 1.0)),
		Vertex(.init( 1.0, -1.0, -1.0), .init(0.0, 1.0)),
		Vertex(.init( 1.0, -1.0,  1.0), .init(0.0, 0.0)),
		Vertex(.init(-1.0, -1.0,  1.0), .init(1.0, 0.0)),
		// Right face
		Vertex(.init( 1.0, -1.0, -1.0), .init(1.0, 0.0)),
		Vertex(.init( 1.0,  1.0, -1.0), .init(1.0, 1.0)),
		Vertex(.init( 1.0,  1.0,  1.0), .init(0.0, 1.0)),
		Vertex(.init( 1.0, -1.0,  1.0), .init(0.0, 0.0)),
		// Left Face
		Vertex(.init(-1.0, -1.0, -1.0), .init(0.0, 0.0)),
		Vertex(.init(-1.0, -1.0,  1.0), .init(1.0, 0.0)),
		Vertex(.init(-1.0,  1.0,  1.0), .init(1.0, 1.0)),
		Vertex(.init(-1.0,  1.0, -1.0), .init(0.0, 1.0)),
	]
	
	static let indices: [UInt16] =
	[
		 0,  1,  2,   2,  3,  0,  // Front
		 4,  5,  6,   6,  7,  4,  // Back
		 8,  9, 10,  10, 11,  8,  // Top
		12, 13, 14,  14, 15, 12,  // Bottom
		16, 17, 18,  18, 19, 16,  // Right
		20, 21, 22,  22, 23, 20   // Left
	]

	var pso: OpaquePointer? = nil
	var vtxBuffer: OpaquePointer? = nil
	var idxBuffer: OpaquePointer? = nil
	var sampler: OpaquePointer? = nil
	var texture: OpaquePointer? = nil
	var projection: matrix_float4x4 = .init(1.0)

	var rot: SIMD3<Float> = .init(repeating: 0.0)

	mutating func `init`(ctx: inout NeHeContext) throws(NeHeError)
	{
		let (vertexShader, fragmentShader) = try ctx.loadShaders(name: "lesson6",
			vertexUniforms: 1, fragmentSamplers: 1)
		defer
		{
			SDL_ReleaseGPUShader(ctx.device, fragmentShader)
			SDL_ReleaseGPUShader(ctx.device, vertexShader)
		}

		let vertexDescriptions: [SDL_GPUVertexBufferDescription] =
		[
			SDL_GPUVertexBufferDescription(
				slot: 0,
				pitch: UInt32(MemoryLayout<Vertex>.stride),
				input_rate: SDL_GPU_VERTEXINPUTRATE_VERTEX,
				instance_step_rate: 0),
		]
		let vertexAttributes: [SDL_GPUVertexAttribute] =
		[
			SDL_GPUVertexAttribute(
				location: 0,
				buffer_slot: 0,
				format: SDL_GPU_VERTEXELEMENTFORMAT_FLOAT3,
				offset: UInt32(MemoryLayout<Vertex>.offset(of: \.position)!)),
			SDL_GPUVertexAttribute(
				location: 1,
				buffer_slot: 0,
				format: SDL_GPU_VERTEXELEMENTFORMAT_FLOAT2,
				offset: UInt32(MemoryLayout<Vertex>.offset(of: \.texcoord)!)),
		]
		let colorTargets: [SDL_GPUColorTargetDescription] =
		[
			SDL_GPUColorTargetDescription(
				format: SDL_GetGPUSwapchainTextureFormat(ctx.device, ctx.window),
				blend_state: SDL_GPUColorTargetBlendState())
		]
		var rasterizerDesc = SDL_GPURasterizerState()
		rasterizerDesc.fill_mode  = SDL_GPU_FILLMODE_FILL
		rasterizerDesc.cull_mode  = SDL_GPU_CULLMODE_NONE
		rasterizerDesc.front_face = SDL_GPU_FRONTFACE_COUNTER_CLOCKWISE
		var depthStencilState = SDL_GPUDepthStencilState()
		depthStencilState.compare_op         = SDL_GPU_COMPAREOP_LESS_OR_EQUAL
		depthStencilState.enable_depth_test  = true
		depthStencilState.enable_depth_write = true
		var targetInfo = SDL_GPUGraphicsPipelineTargetInfo()
		targetInfo.color_target_descriptions = colorTargets.withUnsafeBufferPointer(\.baseAddress!)
		targetInfo.num_color_targets         = UInt32(colorTargets.count)
		targetInfo.depth_stencil_format      = SDL_GPU_TEXTUREFORMAT_D16_UNORM
		targetInfo.has_depth_stencil_target  = true

		var info = SDL_GPUGraphicsPipelineCreateInfo(
			vertex_shader: vertexShader,
			fragment_shader: fragmentShader,
			vertex_input_state: SDL_GPUVertexInputState(
				vertex_buffer_descriptions: vertexDescriptions.withUnsafeBufferPointer(\.baseAddress!),
				num_vertex_buffers: UInt32(vertexDescriptions.count),
				vertex_attributes: vertexAttributes.withUnsafeBufferPointer(\.baseAddress!),
				num_vertex_attributes: UInt32(vertexAttributes.count)),
			primitive_type: SDL_GPU_PRIMITIVETYPE_TRIANGLELIST,
			rasterizer_state: rasterizerDesc,
			multisample_state: SDL_GPUMultisampleState(),
			depth_stencil_state: depthStencilState,
			target_info: targetInfo,
			props: 0)
		guard let pso = SDL_CreateGPUGraphicsPipeline(ctx.device, &info) else
		{
			throw .sdlError("SDL_CreateGPUGraphicsPipeline", String(cString: SDL_GetError()))
		}
		self.pso = pso

		var samplerInfo = SDL_GPUSamplerCreateInfo()
		samplerInfo.min_filter = SDL_GPU_FILTER_LINEAR
		samplerInfo.mag_filter = SDL_GPU_FILTER_LINEAR
		guard let sampler = SDL_CreateGPUSampler(ctx.device, &samplerInfo) else
		{
			throw .sdlError("SDL_CreateGPUSampler", String(cString: SDL_GetError()))
		}
		self.sampler = sampler

		try ctx.copyPass { (pass) throws(NeHeError) in
			self.texture = try pass.createTextureFrom(bmpResource: "NeHe", flip: true)
			self.vtxBuffer = try pass.createBuffer(usage: SDL_GPU_BUFFERUSAGE_VERTEX, Self.vertices[...])
			self.idxBuffer = try pass.createBuffer(usage: SDL_GPU_BUFFERUSAGE_INDEX, Self.indices[...])
		}
	}

	func quit(ctx: NeHeContext)
	{
		SDL_ReleaseGPUBuffer(ctx.device, self.idxBuffer)
		SDL_ReleaseGPUBuffer(ctx.device, self.vtxBuffer)
		SDL_ReleaseGPUTexture(ctx.device, self.texture)
		SDL_ReleaseGPUSampler(ctx.device, self.sampler)
		SDL_ReleaseGPUGraphicsPipeline(ctx.device, self.pso)
	}

	mutating func resize(size: Size<Int32>)
	{
		let aspect = Float(size.width) / Float(max(1, size.height))
		self.projection = .perspective(fovy: 45, aspect: aspect, near: 0.1, far: 100)
	}

	mutating func draw(ctx: inout NeHeContext, cmd: OpaquePointer,
		swapchain: OpaquePointer, swapchainSize: Size<UInt32>) throws(NeHeError)
	{
		var colorInfo = SDL_GPUColorTargetInfo()
		colorInfo.texture     = swapchain
		colorInfo.clear_color = SDL_FColor(r: 0.0, g: 0.0, b: 0.0, a: 0.5)
		colorInfo.load_op     = SDL_GPU_LOADOP_CLEAR
		colorInfo.store_op    = SDL_GPU_STOREOP_STORE

		var depthInfo = SDL_GPUDepthStencilTargetInfo()
		depthInfo.texture          = ctx.depthTexture
		depthInfo.clear_depth      = 1.0
		depthInfo.load_op          = SDL_GPU_LOADOP_CLEAR
		depthInfo.store_op         = SDL_GPU_STOREOP_DONT_CARE
		depthInfo.stencil_load_op  = SDL_GPU_LOADOP_DONT_CARE
		depthInfo.stencil_store_op = SDL_GPU_STOREOP_DONT_CARE
		depthInfo.cycle            = true

		// Begin pass & bind pipeline state
		let pass = SDL_BeginGPURenderPass(cmd, &colorInfo, 1, &depthInfo)
		SDL_BindGPUGraphicsPipeline(pass, self.pso)

		// Bind texture
		var textureBinding = SDL_GPUTextureSamplerBinding(texture: self.texture, sampler: self.sampler)
		SDL_BindGPUFragmentSamplers(pass, 0, &textureBinding, 1)

		// Bind vertex & index buffers
		let vtxBindings = [ SDL_GPUBufferBinding(buffer: self.vtxBuffer, offset: 0) ]
		var idxBinding = SDL_GPUBufferBinding(buffer: self.idxBuffer, offset: 0)
		SDL_BindGPUVertexBuffers(pass, 0,
			vtxBindings.withUnsafeBufferPointer(\.baseAddress!), UInt32(vtxBindings.count))
		SDL_BindGPUIndexBuffer(pass, &idxBinding, SDL_GPU_INDEXELEMENTSIZE_16BIT)

		// Move cube 5 units into the camera and apply some rotations
		var model: simd_float4x4 = .translation(.init(0.0, 0.0, -5.0))
		model.rotate(angle: rot.x, axis: .init(1.0, 0.0, 0.0))
		model.rotate(angle: rot.y, axis: .init(0.0, 1.0, 0.0))
		model.rotate(angle: rot.z, axis: .init(0.0, 0.0, 1.0))

		// Push shader uniforms
		var modelViewProj = self.projection * model
		SDL_PushGPUVertexUniformData(cmd, 0, &modelViewProj, UInt32(MemoryLayout<matrix_float4x4>.size))

		// Draw textured cube
		SDL_DrawGPUIndexedPrimitives(pass, UInt32(Self.indices.count), 1, 0, 0, 0)

		SDL_EndGPURenderPass(pass)

		self.rot += .init(0.3, 0.2, 0.4)
	}
}

@main struct Program: AppRunner
{
	typealias Delegate = Lesson6
	static let config = AppConfig(
		title: "NeHe's Texture Mapping Tutorial",
		width: 640,
		height: 480,
		createDepthBuffer: SDL_GPU_TEXTUREFORMAT_D16_UNORM,
		bundle: Bundle.module)
}

File: src/rust/lesson6.rs
/*
 * SPDX-FileCopyrightText: (C) 2025 a dinosaur
 * SPDX-License-Identifier: Zlib
 */

use nehe::application::config::AppImplementation;
use nehe::application::run;
use nehe::context::NeHeContext;
use nehe::error::NeHeError;
use nehe::matrix::Mtx;
use sdl3_sys::gpu::*;
use sdl3_sys::pixels::SDL_FColor;
use std::cmp::max;
use std::ffi::c_void;
use std::mem::offset_of;
use std::process::ExitCode;
use std::ptr::{addr_of, null_mut};

#[repr(C)]
struct Vertex
{
	x: f32, y: f32, z: f32,
	u: f32, v: f32,
}

const VERTICES: &'static [Vertex] =
&[
	// Front Face
	Vertex { x: -1.0, y: -1.0, z:  1.0, u: 0.0, v: 0.0 },
	Vertex { x:  1.0, y: -1.0, z:  1.0, u: 1.0, v: 0.0 },
	Vertex { x:  1.0, y:  1.0, z:  1.0, u: 1.0, v: 1.0 },
	Vertex { x: -1.0, y:  1.0, z:  1.0, u: 0.0, v: 1.0 },
	// Back Face
	Vertex { x: -1.0, y: -1.0, z: -1.0, u: 1.0, v: 0.0 },
	Vertex { x: -1.0, y:  1.0, z: -1.0, u: 1.0, v: 1.0 },
	Vertex { x:  1.0, y:  1.0, z: -1.0, u: 0.0, v: 1.0 },
	Vertex { x:  1.0, y: -1.0, z: -1.0, u: 0.0, v: 0.0 },
	// Top Face
	Vertex { x: -1.0, y:  1.0, z: -1.0, u: 0.0, v: 1.0 },
	Vertex { x: -1.0, y:  1.0, z:  1.0, u: 0.0, v: 0.0 },
	Vertex { x:  1.0, y:  1.0, z:  1.0, u: 1.0, v: 0.0 },
	Vertex { x:  1.0, y:  1.0, z: -1.0, u: 1.0, v: 1.0 },
	// Bottom Face
	Vertex { x: -1.0, y: -1.0, z: -1.0, u: 1.0, v: 1.0 },
	Vertex { x:  1.0, y: -1.0, z: -1.0, u: 0.0, v: 1.0 },
	Vertex { x:  1.0, y: -1.0, z:  1.0, u: 0.0, v: 0.0 },
	Vertex { x: -1.0, y: -1.0, z:  1.0, u: 1.0, v: 0.0 },
	// Right face
	Vertex { x:  1.0, y: -1.0, z: -1.0, u: 1.0, v: 0.0 },
	Vertex { x:  1.0, y:  1.0, z: -1.0, u: 1.0, v: 1.0 },
	Vertex { x:  1.0, y:  1.0, z:  1.0, u: 0.0, v: 1.0 },
	Vertex { x:  1.0, y: -1.0, z:  1.0, u: 0.0, v: 0.0 },
	// Left Face
	Vertex { x: -1.0, y: -1.0, z: -1.0, u: 0.0, v: 0.0 },
	Vertex { x: -1.0, y: -1.0, z:  1.0, u: 1.0, v: 0.0 },
	Vertex { x: -1.0, y:  1.0, z:  1.0, u: 1.0, v: 1.0 },
	Vertex { x: -1.0, y:  1.0, z: -1.0, u: 0.0, v: 1.0 },
];

const INDICES: &'static [u16] =
&[
	 0,  1,  2,   2,  3,  0,  // Front
	 4,  5,  6,   6,  7,  4,  // Back
	 8,  9, 10,  10, 11,  8,  // Top
	12, 13, 14,  14, 15, 12,  // Bottom
	16, 17, 18,  18, 19, 16,  // Right
	20, 21, 22,  22, 23, 20   // Left
];

struct Lesson6
{
	pso: *mut SDL_GPUGraphicsPipeline,
	vtx_buffer: *mut SDL_GPUBuffer,
	idx_buffer: *mut SDL_GPUBuffer,
	sampler: *mut SDL_GPUSampler,
	texture: *mut SDL_GPUTexture,
	projection: Mtx,

	rot: (f32, f32, f32),
}

//FIXME: remove when `raw_ptr_default`
impl Default for Lesson6
{
	fn default() -> Self
	{
		Self
		{
			pso: null_mut(),
			vtx_buffer: null_mut(),
			idx_buffer: null_mut(),
			sampler: null_mut(),
			texture: null_mut(),
			projection: Mtx::IDENTITY,

			rot: (0.0, 0.0, 0.0),
		}
	}
}

impl AppImplementation for Lesson6
{
	const TITLE: &'static str = "NeHe's Texture Mapping Tutorial";
	const WIDTH: i32 = 640;
	const HEIGHT: i32 = 480;
	const CREATE_DEPTH_BUFFER: SDL_GPUTextureFormat = SDL_GPU_TEXTUREFORMAT_D16_UNORM;

	fn init(&mut self, ctx: &NeHeContext) -> Result<(), NeHeError>
	{
		let (vertex_shader, fragment_shader) = ctx.load_shaders("lesson6", 1, 0, 1)?;

		const VERTEX_DESCRIPTIONS: &'static [SDL_GPUVertexBufferDescription] =
		&[
			SDL_GPUVertexBufferDescription
			{
				slot: 0,
				pitch: size_of::<Vertex>() as u32,
				input_rate: SDL_GPU_VERTEXINPUTRATE_VERTEX,
				instance_step_rate: 0,
			},
		];
		const VERTEX_ATTRIBS: &'static [SDL_GPUVertexAttribute] =
		&[
			SDL_GPUVertexAttribute
			{
				location: 0,
				buffer_slot: 0,
				format: SDL_GPU_VERTEXELEMENTFORMAT_FLOAT3,
				offset: offset_of!(Vertex, x) as u32,
			},
			SDL_GPUVertexAttribute
			{
				location: 1,
				buffer_slot: 0,
				format: SDL_GPU_VERTEXELEMENTFORMAT_FLOAT2,
				offset: offset_of!(Vertex, u) as u32,
			},
		];

		let mut pso_info = SDL_GPUGraphicsPipelineCreateInfo::default();
		pso_info.vertex_shader   = vertex_shader;
		pso_info.fragment_shader = fragment_shader;
		pso_info.primitive_type  = SDL_GPU_PRIMITIVETYPE_TRIANGLELIST;
		pso_info.vertex_input_state = SDL_GPUVertexInputState
		{
			vertex_buffer_descriptions: VERTEX_DESCRIPTIONS.as_ptr(),
			num_vertex_buffers: VERTEX_DESCRIPTIONS.len() as u32,
			vertex_attributes: VERTEX_ATTRIBS.as_ptr(),
			num_vertex_attributes: VERTEX_ATTRIBS.len() as u32,
		};
		pso_info.rasterizer_state.fill_mode  = SDL_GPU_FILLMODE_FILL;
		pso_info.rasterizer_state.cull_mode  = SDL_GPU_CULLMODE_NONE;
		pso_info.rasterizer_state.front_face = SDL_GPU_FRONTFACE_COUNTER_CLOCKWISE;
		let color_targets =
		[
			SDL_GPUColorTargetDescription
			{
				format: unsafe { SDL_GetGPUSwapchainTextureFormat(ctx.device, ctx.window) },
				blend_state: SDL_GPUColorTargetBlendState::default(),
			}
		];
		pso_info.target_info.color_target_descriptions = color_targets.as_ptr();
		pso_info.target_info.num_color_targets         = color_targets.len() as u32;
		pso_info.target_info.depth_stencil_format      = SDL_GPU_TEXTUREFORMAT_D16_UNORM;
		pso_info.target_info.has_depth_stencil_target  = true;

		pso_info.depth_stencil_state.compare_op         = SDL_GPU_COMPAREOP_LESS_OR_EQUAL;
		pso_info.depth_stencil_state.enable_depth_test  = true;
		pso_info.depth_stencil_state.enable_depth_write = true;

		self.pso = unsafe { SDL_CreateGPUGraphicsPipeline(ctx.device, &pso_info) };
		unsafe
		{
			SDL_ReleaseGPUShader(ctx.device, fragment_shader);
			SDL_ReleaseGPUShader(ctx.device, vertex_shader);
		}
		if self.pso.is_null()
		{
			return Err(NeHeError::sdl("SDL_CreateGPUGraphicsPipeline"));
		}

		// Create texture sampler
		let mut sampler_info = SDL_GPUSamplerCreateInfo::default();
		sampler_info.min_filter = SDL_GPU_FILTER_LINEAR;
		sampler_info.mag_filter = SDL_GPU_FILTER_LINEAR;
		self.sampler = unsafe { SDL_CreateGPUSampler(ctx.device, &sampler_info).as_mut() }
			.ok_or(NeHeError::sdl("SDL_CreateGPUSampler"))?;

		ctx.copy_pass(|pass|
		{
			self.texture = pass.load_texture("Data/NeHe.bmp", true, false)?;
			self.vtx_buffer = pass.create_buffer(SDL_GPU_BUFFERUSAGE_VERTEX, VERTICES)?;
			self.idx_buffer = pass.create_buffer(SDL_GPU_BUFFERUSAGE_INDEX, INDICES)?;
			Ok(())
		})
	}

	fn quit(&mut self, ctx: &NeHeContext)
	{
		unsafe
		{
			SDL_ReleaseGPUBuffer(ctx.device, self.idx_buffer);
			SDL_ReleaseGPUBuffer(ctx.device, self.vtx_buffer);
			SDL_ReleaseGPUTexture(ctx.device, self.texture);
			SDL_ReleaseGPUSampler(ctx.device, self.sampler);
			SDL_ReleaseGPUGraphicsPipeline(ctx.device, self.pso);
		}
	}

	fn resize(&mut self, _ctx: &NeHeContext, width: i32, height: i32)
	{
		let aspect = width as f32 / max(height, 1) as f32;
		self.projection = Mtx::perspective(45.0, aspect, 0.1, 100.0);
	}

	fn draw(&mut self, ctx: &NeHeContext, cmd: *mut SDL_GPUCommandBuffer, swapchain: *mut SDL_GPUTexture)
	{
		let mut color_info = SDL_GPUColorTargetInfo::default();
		color_info.texture     = swapchain;
		color_info.clear_color = SDL_FColor { r: 0.0, g: 0.0, b: 0.0, a: 0.5 };
		color_info.load_op     = SDL_GPU_LOADOP_CLEAR;
		color_info.store_op    = SDL_GPU_STOREOP_STORE;

		let mut depth_info = SDL_GPUDepthStencilTargetInfo::default();
		depth_info.texture          = ctx.depth_texture;
		depth_info.clear_depth      = 1.0;
		depth_info.load_op          = SDL_GPU_LOADOP_CLEAR;
		depth_info.store_op         = SDL_GPU_STOREOP_DONT_CARE;
		depth_info.stencil_load_op  = SDL_GPU_LOADOP_DONT_CARE;
		depth_info.stencil_store_op = SDL_GPU_STOREOP_DONT_CARE;
		depth_info.cycle            = true;

		unsafe
		{
			// Begin pass & bind pipeline state
			let pass = SDL_BeginGPURenderPass(cmd, &color_info, 1, &depth_info);
			SDL_BindGPUGraphicsPipeline(pass, self.pso);

			// Bind texture
			let texture_binding = SDL_GPUTextureSamplerBinding { texture: self.texture, sampler: self.sampler };
			SDL_BindGPUFragmentSamplers(pass, 0, &texture_binding, 1);

			// Bind vertex & index buffers
			let vtx_binding = SDL_GPUBufferBinding { buffer: self.vtx_buffer, offset: 0 };
			let idx_binding = SDL_GPUBufferBinding { buffer: self.idx_buffer, offset: 0 };
			SDL_BindGPUVertexBuffers(pass, 0, &vtx_binding, 1);
			SDL_BindGPUIndexBuffer(pass, &idx_binding, SDL_GPU_INDEXELEMENTSIZE_16BIT);

			// Move cube 5 units into the screen and apply some rotations
			let mut model = Mtx::translation(0.0, 0.0, -5.0);
			model.rotate(self.rot.0, 1.0, 0.0, 0.0);
			model.rotate(self.rot.1, 0.0, 1.0, 0.0);
			model.rotate(self.rot.2, 0.0, 0.0, 1.0);

			// Push shader uniforms
			let model_view_proj = self.projection * model;
			SDL_PushGPUVertexUniformData(cmd, 0,
				addr_of!(model_view_proj) as *const c_void,
				size_of::<Mtx>() as u32);

			// Draw the textured cube
			SDL_DrawGPUIndexedPrimitives(pass, INDICES.len() as u32, 1, 0, 0, 0);

			SDL_EndGPURenderPass(pass);
		}

		self.rot.0 += 0.3;
		self.rot.1 += 0.2;
		self.rot.2 += 0.4;
	}
}

pub fn main() -> Result<ExitCode, Box<dyn std::error::Error>>
{
	run::<Lesson6>()?;
	Ok(ExitCode::SUCCESS)
}
